bittide/src/Bittide/Transceiver/ResetManager.hs

```diff
@@ -0,0 +1,186 @@
+  -- times out. Depending on the value of 'ResetStat's 'rxRetries' it will
+  -- either reset both the receive and the transmit side, or just the receive
+  -- side. If all is well though, move on to 'Monitor'.
+  | Monitor Statistics
+  -- ^ Wait till the end of the universe, or until a link goes down - whichever
+  -- comes first. In case of the latter, the state machine moves to 'StartTx'.
+  deriving (Generic, NFDataX)
+
+-- | Reset manager for transceivers: see 'State' for more information on
+-- this state machine. See 'Statistics' for information on what debug values
+-- are exported.
+resetManager ::
+  forall dom .
+  KnownDomain dom =>
+  Config ->
+  Clock dom ->
+  Reset dom ->
+  "tx_init_done" ::: Signal dom Bool ->
+  "rx_init_done" ::: Signal dom Bool ->
+  "rx_data_good" ::: Signal dom Bool ->
+  ( "reset_all_out" ::: Reset dom
+  , "reset_rx"  ::: Reset dom
+  , "stats" ::: Signal dom Statistics
+  )
+resetManager config clk rst tx_init_done rx_init_done rx_data_good =
+  ( unsafeFromActiveHigh reset_all_out_sig
+  , unsafeFromActiveHigh reset_rx_sig
+  , statistics
+  )
+ where
+  (reset_all_out_sig, reset_rx_sig, statistics) =
+    mooreB
+      clk rst enableGen
+      update
+      extractOutput
+      initSt
+      ( tx_init_done
+      , rx_init_done
+      , rx_data_good
+      )
+
+  initSt :: State dom
+  initSt = StartTx (Statistics
+    { rxRetries=0
+    , rxFullRetries=0
+    , txRetries=0
+    , failAfterUps=0
+    })
+
+  update :: State dom -> (Bool, Bool, Bool) -> State dom
+  update st (tx_done, rx_done, rx_good) =
+    case st of
+      -- Reset everything:
+      StartTx stats -> TxWait stats minBound
+
+      -- Wait for transceiver to indicate it is done
+      TxWait stats@Statistics{txRetries} cntr@(ms, _)
+        | tx_done                  -> RxWait stats minBound
+        | ms == config.txTimeoutMs -> StartTx stats{txRetries=satSucc SatBound txRetries}
+        | otherwise                -> TxWait stats (countSucc cntr)
+
+      -- Reset receive side logic
+      StartRx stats -> RxWait stats minBound
+
+      -- Wait for a reliable incoming link. This can fail in multiple ways, see
+      -- 'RxWait'.
+      RxWait stats@Statistics{rxRetries, rxFullRetries} cntr@(ms, _)
+        | rx_done && rx_good ->
+          Monitor stats
+
+        | ms == config.rxTimeoutMs && rxRetries >= config.rxRetries ->
+          StartTx stats{rxFullRetries=satSucc SatBound rxFullRetries}
+
+        | ms == config.rxTimeoutMs ->
+          StartRx stats{rxRetries=satSucc SatBound rxRetries}
+
+        | otherwise ->
+          RxWait stats (countSucc cntr)
+
+      -- Monitor link. Move all the way back to 'StartTx' if the link goes down
+      -- for some reason.
+      Monitor stats@Statistics{failAfterUps}
+        | rx_done && rx_good -> Monitor stats
+        | otherwise -> StartTx stats{failAfterUps=satSucc SatBound failAfterUps}
+
+  extractOutput st = case st of
```

Comment:
With externalizing `Statistics` from `State` and deriving `Eq`, we could for example reduce this to
```haskell
extractOutput (st, stats) = (st == StartTx, st == StartRx, stats)
```
making it much more readable IMO.

Reply:
Agreed, done